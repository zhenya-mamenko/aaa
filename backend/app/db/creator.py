import csv
import os
import sqlite3
import sys

from argparse import ArgumentParser

from app.config import settings


def create_schema(cursor, schema_files: list) -> None:
    """
    Create the database schema from the provided schema files.

    Args:
        cursor: The database cursor.
        schema_files (list): List of schema file paths.

    Returns:
        None
    """

    for schema_file in schema_files:
        with open(schema_file, "r") as f:
            schema = f.read()
            cursor.executescript(schema)


def import_data(cursor, data_files: list, delimeter: ord) -> None:
    """
    Import data into the database from the provided data files.

    Args:
        cursor: The database cursor.
        data_files (list): List of data file paths.
        delimeter (ord): The delimiter used in the data files.

    Returns:
        None
    """

    for data_file in data_files:
        table = os.path.splitext(os.path.basename(data_file))[0]
        with open(data_file, "r", encoding="utf-8") as f:
            reader = csv.DictReader(f, delimiter=chr(delimeter))
            if not reader.fieldnames: # pragma: no cover
                continue
            cols = ", ".join(reader.fieldnames)
            template = ", ".join(["?"] * len(reader.fieldnames))
            sql = f"INSERT INTO {table} ({cols}) VALUES ({template})"
            rows = [tuple(row.values()) for row in reader]
            cursor.executemany(sql, rows)


def parse_args(sys_args: list[str]) -> dict:
    """
    Parse command-line arguments.

    Args:
        sys_args (list[str]): List of command-line arguments.

    Returns:
        dict: Parsed arguments as a dictionary.
    """

    parser = ArgumentParser(
        prog="python -m app.db.creator",
        description=f"Create the schema for the {settings.app_name} database and import data if provided",
        epilog=f"See more on {settings.github_link}",
        allow_abbrev=False,
    )
    parser.add_argument("db_name", help="database filename", default=settings.database_path)
    parser.add_argument("-s", "--schema", nargs="*", help="filenames for db schema")
    parser.add_argument("-f", "--data", nargs="*", help="filenames for importing data")
    parser.add_argument("-d", type=ord, help="char for delimeter in data files", default="|")
    args = parser.parse_args(sys_args)

    return dict(db_name=args.db_name, schema=args.schema, data=args.data, delimeter=args.d)


def run(
    *,
    db_name: str | None = None,
    db_conn: sqlite3.Connection | None = None,
    schema: list = [],
    data: list = [],
    delimeter: ord = ord("|"),
) -> None:
    """
    Run the database creation and data import process.

    Args:
        db_name (str | None): The name of the database file.
        db_conn (sqlite3.Connection | None): The database connection.
        schema (list): List of schema file paths.
        data (list): List of data file paths.
        delimeter (ord): The delimiter used in the data files.

    Returns:
        None

    Raises:
        ValueError: If neither db_name nor db_conn is provided.
    """

    if not db_name and not db_conn:
        raise ValueError("Either db_name or db_conn must be provided")
    if not db_conn:
        conn = sqlite3.connect(db_name)
    else:
        conn = db_conn
    cursor = conn.cursor()

    if schema:
        create_schema(cursor, schema)
    if data:
        import_data(cursor, data, delimeter)

    conn.commit()
    if not db_conn:
        conn.close()


if __name__ == "__main__":
    args = parse_args(sys_args=sys.argv[1:])

    run(**args)
