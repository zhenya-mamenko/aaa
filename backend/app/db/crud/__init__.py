# ruff: noqa: F403
from app.db.crud.assets import *
from app.db.crud.categories import *
from app.db.crud.classes import *
from app.db.crud.config import *
from app.db.crud.crud import *
from app.db.crud.structures import *
