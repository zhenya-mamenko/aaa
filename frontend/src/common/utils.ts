import { type AppElement } from "@/types/common";
import { elements } from "@/common/const";


export const isPlainObject: Function = (o: any) => Boolean(
  o && o.constructor && o.constructor.prototype && o.constructor.prototype.hasOwnProperty("isPrototypeOf")
);


export const flattenObject: Function = (obj: any, keys = []) => {
  return Object.keys(obj).reduce((acc: any, key: any) => {
    return Object.assign(acc, isPlainObject(obj[key])
      ? flattenObject(obj[key], keys.concat(key))
      : { [keys.concat(key).join(".")]: obj[key] }
    )
  }, {})
}


export const loadParams: Function = (key: string) => {
  const savedParams = localStorage.getItem(key)
  if (savedParams) {
    return JSON.parse(savedParams)
  }
  return null
}


export const saveParams: Function = (key: string, params: any) => {
  localStorage.setItem(key, JSON.stringify(params))
}


export const deleteParams: Function = (key: string) => {
  localStorage.removeItem(key)
}


export const removeReactivity: Function = (o: any) => {
  return JSON.parse(JSON.stringify(o))
}

export const findElement: Function = (key: string): AppElement | undefined => {
  return elements.find((element: AppElement) => element.key === key);
}


// String.formatUnicorn polyfill from https://stackoverflow.com/a/18234317
export const format: Function = function(str: string, ...rest: any[]): string {
    if (rest.length) {
      var t = typeof rest[0];
      var key;
      var args = ("string" === t || "number" === t) ? Array.prototype.slice.call(rest) : rest[0];
      for (key in args) {
        str = str.replace(new RegExp("\\{" + key + "\\}", "gi"), args[key]);
      }
    }
    return str;
  };


export const parseAxiosError: Function = (error: any): string | any[] => {
  if (error.response?.data?.detail !== undefined) {
    let message = "";
    const detail = error.response.data.detail;
    if (typeof detail === "string" && detail.length > 0) {
      message = `messages.backend.${detail.split(":")[0].replace(/ /g, "_")}`;
    }
    return [
      "messages.backend.request_error",
      error.response.status,
      message,
    ];
  } else if (!error.response) {
    return "messages.backend.no_reponse";
  } else {
    return error.message || "messages.backend.unknown_error";
  }
}


export const calculateReplenishment: Function = (
  list: any[],
  summa: number,
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(summa * item[percentileKey] / 1000);
  });
}


export const calculateRebalancing: Function = (
  list: any[],
  summa: number,
  amountKey: string = "amount",
  percentileKey: string = "structure_percentile",
  resultKey: string = "calculated"
): void => {
  const total = list.reduce((acc, item) => acc + item[amountKey], 0) + summa * 100;
  list.forEach(item => {
    (item as any)[resultKey] = Math.round(((total * item[percentileKey] / 1000) - item[amountKey]) / 100);
  });
}
