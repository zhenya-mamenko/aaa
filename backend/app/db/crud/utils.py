import json
import re

from pydantic import BaseModel as _BaseModel

from app.db.sqlite import (
    execute_query as _execute_query,
    get_database_conn as _get_database_conn,
    read_query as _read_query,
)


def create_parameters_from_model(
    model: _BaseModel,
    /,
    exclude_fields: tuple[str] | None = None,
    only_fields: tuple[str] | None = None,
    get_max_data: tuple | None = None,
) -> tuple:
    """
    Create SQL parameters from a Pydantic model.

    Args:
        model (_BaseModel): The Pydantic model instance.
        exclude_fields (tuple[str] | None, optional): Fields to exclude from the model. Defaults to None.
        only_fields (tuple[str] | None, optional): Fields to include from the model. Defaults to None.
        get_max_data (tuple | None, optional): Tuple containing table name and field to get max value. Defaults to None.

    Returns:
        tuple: A tuple containing the SQL template, fields, parameters, and optionally a max query.
    """

    exclude_fields = exclude_fields or ()
    model_dump = {
        k: json.dumps(v) if type(v).__name__ in ["list", "dict"] else v for k, v in model.model_dump().items()
    }

    if only_fields:
        fields = [field for field in only_fields if field in model_dump]
        if len(fields) != len(only_fields):
            raise ValueError(f"Fields={set(fields) ^ set(only_fields)} are not in the model.")
    else:
        fields = [field for field, value in model_dump.items() if field not in exclude_fields and value is not None]
    parameters = tuple([model_dump[field] for field in fields])
    template = ", ".join(["?" for _ in parameters])

    if get_max_data:
        table_name, field = get_max_data
        if field in fields:
            raise ValueError(f"Field {field} must be excluded from the model.")
        fields.append(field)
        template += f", (SELECT COALESCE(MAX({field}), 0) + 1 FROM {table_name})"
        max_query = f"SELECT MAX({field}) as {field} FROM {table_name};"
        return template, fields, parameters, max_query

    return template, fields, parameters


def insert_data_from_model(
    *, table_name: str, id_field: tuple | str, model: _BaseModel, auto_increment_id: bool = True
) -> int | str | None:
    """
    Insert data into a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to insert data into.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.
        auto_increment_id (bool, optional): Whether the ID is auto-incremented. Defaults to True.

    Returns:
        int | str | None: The ID of the inserted row if auto_increment_id is True, otherwise None.
    """

    if auto_increment_id:
        if isinstance(id_field, tuple):
            raise ValueError("auto_increment_id must be False if id_field is a tuple.")
        template, fields, parameters, q = create_parameters_from_model(model, get_max_data=(table_name, id_field))
    else:
        template, fields, parameters = create_parameters_from_model(model)

    with _get_database_conn() as conn:
        query = f"""
            INSERT INTO {table_name} (
                {", ".join(fields)}
            ) VALUES (
                {template}
            );
        """
        _execute_query(query, parameters=parameters, conn=conn)
        if auto_increment_id:
            id = _read_query(q, conn=conn)[0][id_field]
            return id

    return None


def update_data_from_model(*, table_name: str, id_field: tuple | str, model: _BaseModel) -> bool:
    """
    Update data in a table from a Pydantic model.

    Args:
        table_name (str): The name of the table to update data in.
        id_field (tuple | str): The field(s) representing the ID.
        model (_BaseModel): The Pydantic model instance.

    Returns:
        bool: True if the update was successful, False otherwise.
    """

    if isinstance(id_field, str):
        id_field = (id_field,)

    _, fields, parameters = create_parameters_from_model(model, exclude_fields=id_field)

    query = f"""
        UPDATE {table_name}
        SET
            {", ".join([f"{field} = ?" for field in fields])}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    parameters += tuple([getattr(model, f) for f in id_field])
    result = _execute_query(query, parameters=parameters)
    return result == 1


def delete_data_by_id(*, table_name: str, id_field: tuple | str, id_value: int | str | tuple) -> bool:
    """
    Delete data from a table by ID.

    Args:
        table_name (str): The name of the table to delete data from.
        id_field (tuple | str): The field(s) representing the ID.
        id_value (int | str | tuple): The value(s) of the ID to delete.

    Returns:
        bool: True if the deletion was successful, False otherwise.
    """

    if not isinstance(id_field, tuple):
        id_field=(id_field, )
    query = f"""
        DELETE FROM {table_name}
        WHERE
            {' AND '.join([f'{f} = ?' for f in id_field])};
    """
    if not isinstance(id_value, tuple):
        id_value=(id_value, )
    result = _execute_query(query, parameters=id_value)

    return result == 1


def extract_create_stmts_objects(sql_file_path: str, object_type: str) -> list[str]:
    """
    Extract the CREATE statements objects from a SQL file.

    Args:
        sql_file_path (str): The path to the SQL file.
        object_type (str): The type of objects to extract from the SQL file.

    Returns:
        list: A list of dictionaries containing the objects.
    """

    try:
        with open(sql_file_path, "r") as f:
            sql = f.read()
    except FileNotFoundError:
        raise FileNotFoundError(f"File not found: {sql_file_path}")

    re_template = fr"CREATE\s+{object_type}\s+(IF\s+NOT\s+EXISTS\s+)*(\S+)"
    pattern = re.compile(re_template)
    result = []
    for m in pattern.finditer(sql):
        result.append(m.group(2))

    return result
