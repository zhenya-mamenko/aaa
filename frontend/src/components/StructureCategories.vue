<template>
  <v-data-table data-testid="structure-categories-data-table"
    class="h-100"
    :headers="headers"
    hide-default-footer
    :items="categories"
    items-per-page="100"
    :loading="isLoading"
    :loading-text="t('messages.loading')"
    sort-asc-icon="mdi-triangle-small-up"
    sort-desc-icon="mdi-triangle-small-down"
    :sticky="true"
  >
    <template v-slot:top>
      <v-toolbar flat>
        <v-icon
          class="ms-3 me-3"
          size="large"
        >
          {{ icon }}
        </v-icon>
        <v-toolbar-title
          class="d-none d-md-block ms-0"
        >
          {{ title }}
        </v-toolbar-title>
        <v-responsive class="w-25">
          <v-select
            class="me-2"
            data-testid="structure-categories-structure-select"
            density="compact"
            hide-details
            :items="structures"
            :prepend-inner-icon="structureIcon"
            ref="structureSelect"
            v-model="structure_id"
            @update:modelValue="loadCategories"
          />
        </v-responsive>
        <v-divider class="ms-3 me-3" vertical />
        <v-btn
          class="me-2"
          data-testid="structure-categories-add-button"
          color="primary-darken-3"
          :disabled="structures.length === 0"
          @click="createItem"
        >
          {{ t("buttons.add") }}
        </v-btn>
    </v-toolbar>
  </template>
  <template v-slot:item.actions="{ item }">
      <v-icon class="me-2" size="small" @click="editItem(item)">
        mdi-pencil
      </v-icon>
      <v-icon size="small" @click="confirmDelete(item)">
        mdi-delete
      </v-icon>
    </template>
    <template v-slot:bottom>
      <v-toolbar flat>
        <v-btn
          class="ms-3"
          data-testid="structure-categories-set-as-current-button"
          color="primary-darken-3"
          :disabled="is_current || structureSum !== 100"
          @click="setAsDefault"
        >
          {{ t("structures.set_as_current") }}
        </v-btn>
        <v-divider class="ms-3 me-3" vertical />
        <v-spacer/>
        <v-responsive class="text-right text-h6 me-3">
          {{ structureSum }}%
          <v-icon
            v-if="structureSum !== 100"
            class="ms-1"
            color="error"
            icon="mdi-alert"
            size="small"
          />
        </v-responsive>
      </v-toolbar>
    </template>
  </v-data-table>
  <form-dialog
    :is-saving="isSaving"
    :opened="dialog.opened"
    :params="dialog.params"
    :show-reset-button="false"
    :title="dialog.title"
    @close="close"
    @save="save"
  />
  <delete-confirmation-dialog
    :opened="confirmationDialog.opened"
    :text="confirmationDialog.text"
    @close="confirmationDialog.opened = false"
    @delete="deleteItem(confirmationDialog.id)"
  />
  <v-snackbar data-testid="structure-categories-snackbar"
    :timeout="info.timeout || 2000"
    :color="info.color || 'success'"
    vertical
    v-model="info.show"
  >
    {{ info.text }}
  </v-snackbar>
</template>

<script lang="ts">
import { defineComponent } from 'vue';
import { state } from "@/common/state";
import { t } from "@/common/locale";
import { findElement, loadParams, parseAxiosError } from "@/common/utils";

import ds from "@/services/StructuresDataService";
import dsCategories from "@/services/CategoriesDataService";
import { type StructureCategory, type StructureCategoryResponse, type StructureResponse } from "@/types/Structures";
import { type Category } from "@/types/Categories";


const pageKey = "structure_categories";

export default defineComponent({
  computed: {
    headers() {
      const columns = this.params.columns;
      if (!columns) return [];
      const result = columns
        .filter(column => column.visible)
        .map(column => {
          const header: any = this.headersRaw.find(header => header.key === column.key);
          header.title = t(column.title);
          header.align = header.align as "start" | "center" | "end";
          return header;
        });
      result.push({ title: t("actions"), sortable: false, align: "center", maxWidth: "80px", width: "80px", value: "actions", });
      return result;
    },
    params() {
      return state.settings[pageKey];
    },
    t() {
      return t;
    },
  },
  data() {
    return {
      structures: [] as any[],
      structure_id: "" as string | number,
      categories: [] as StructureCategoryResponse[],
      confirmationDialog: {
        id: 0 as number,
        opened: false,
        text: "",
      },
      dialog: {
        id: 0 as number,
        item: {} as any,
        opened: false,
        params: {},
        title: "",
      },
      headersRaw: [
        {
          key: "class",
          value: (item: any) => item.class_name,
        },
        {
          key: "category",
          value: (item: any) => item.category_name,
        },
        {
          align: "end",
          key: "target_percent",
          maxWidth: "130px",
          noWrap: true,
          sortRaw(a: StructureCategoryResponse, b: StructureCategoryResponse): number {
            return a.percentile > b.percentile ? 1 : (a.percentile < b.percentile ? -1 : 0);
          },
          value: (item: any) => item.out_percentile,
          width: "100px",
        },
      ],
      icon: "",
      info: {
        color: "success",
        show: false,
        text: "",
        timeout: 2000,
      } as Partial<{ color: string, show: boolean, text: string, timeout: number }>,
      is_current: false,
      isLoading: false,
      isSaving: false,
      structureIcon: "mdi-list-box",
      structureSum: 0,
      title: "",
    }
  },
  methods: {
    async add(item: StructureCategory) {
      try {
        await ds.create(item);
        this.loadCategories();
        this.close();
        this.info = {
          show: true,
          text: t("messages.added"),
        }
      } catch (error: any) {
        const res = parseAxiosError(error);
        this.info = {
          color: "error",
          show: true,
          timeout: 3000,
          text: Array.isArray(res) ? t(res[0], ...res.slice(1)) : t(res),
        }
        console.error(error);
      }
    },
    close() {
      this.dialog.opened = false;
      this.dialog.item = {};
      this.dialog.id = 0;
    },
    async createItem() {
      this.dialog.title = t("buttons.add") + " " + t("structures.element").toLowerCase();
      this.dialog.params = await this.makeParams();
      this.dialog.opened = true;
    },
    async confirmDelete(item: StructureCategoryResponse) {
      this.confirmationDialog.text = t(
        "confirmations.delete_text", item.category_name
      );
      this.confirmationDialog.id = item.category_id;
      this.confirmationDialog.opened = true;
    },
    async deleteItem(id: number) {
      try {
        this.confirmationDialog.opened = false;
        await ds.delete(this.structure_id as number, id);
        this.loadCategories()
        this.info = {
          show: true,
          text: t("messages.deleted"),
        }
      } catch (error: any) {
        const res = parseAxiosError(error);
        this.info = {
          color: "error",
          show: true,
          timeout: 3000,
          text: Array.isArray(res) ? t(res[0], ...res.slice(1)) : t(res),
        }
        console.error(error);
      }
    },
    async editItem(item: StructureCategory) {
      this.dialog.item = item;
      this.dialog.id = item.category_id;
      this.dialog.params = await this.makeParams();
      this.dialog.title = t("buttons.edit") + " " + t("structures.element").toLowerCase();
      this.dialog.opened = true;
    },
    async loadData() {
      this.isLoading = true;
      this.title = t(findElement(pageKey).title);
      this.icon = findElement(pageKey).icon;
      if (!ds) return;
      const structures = await ds.getAll() as StructureResponse[];
      this.structures = structures.map((item: StructureResponse) => ({
        value: item.structure_id,
        title: `${item.structure_name}${item.is_current ? ` âœ”` : ""}`,
        is_current: item.is_current,
      }));
      if (this.structures.length !== 0) {
        this.structure_id = structures.find((item: any) => Boolean(item.is_current))?.structure_id || structures[0].structure_id;
        await this.loadCategories();
      }
      this.isLoading = false;
    },
    async loadCategories() {
      if (!ds || !this.structure_id) return;
      this.isLoading = true;
      (this.$refs.structureSelect as HTMLInputElement)?.blur();
      this.is_current = this.structures.find((item: any) => item.value === this.structure_id)?.is_current;
      const categories = await ds.getAll(this.structure_id as number) as StructureCategoryResponse[];
      this.categories = categories;
      this.structureSum = categories.reduce((acc: number, item: StructureCategoryResponse) => acc + item.percentile, 0) / 10;
      this.isLoading = false;
    },
    async makeParams() {
      return {
        set: [
          {
            cols: 9,
            key: "category_id",
            rules: [
              "required",
            ],
            title: t("categories.1"),
            type: "autocomplete",
            value: (await dsCategories.getAll())
              .map((a: Category) => ({
                value: a.category_id,
                title: a.category_name,
                selected: this.dialog.id && this.dialog.item.category_id === a.category_id,
              })),
          },
          {
            key: "percentile",
            cols: 3,
            rules: [
              "required",
              "number",
              ["value_range", 1, 99],
            ],
            title: t("portfolio.target_percent"),
            type: "number",
            value: !!this.dialog.id ? (this.dialog.item.percentile / 10) : "",
          },
        ],
      };
    },
    save(data: any) {
      this.isSaving = true;
      const result = {
        category_id: this.dialog.item.category_id,
        percentile: this.dialog.item.percentile,
        structure_id: this.dialog.item.structure_id,
      } as any;
      const gv = (a: any) => Array.isArray(a.value) ? a.value.find((item: any) => item.selected).value ?? null : a.value;

      Object.entries(data).forEach(([key, value]) => {
        if (Array.isArray(value)) {
          for (const item of value) {
            result[item.key] = gv(item);
          }
        } else {
          result[key] = gv(value);
        }
      });
      result.percentile = Math.round(result.percentile * 10);
      result.structure_id = this.structure_id;
      if (this.dialog.id) {
        this.update(result);
      } else {
        this.add(result);
      }
      this.isSaving = false;
    },
    async setAsDefault() {
      const s = this.categories.reduce((acc: number, item: StructureCategoryResponse) => acc + item.percentile, 0);
      if (s !== 1000) {
        this.info = {
          color: "error",
          show: true,
          text: t("messages.less_100_percents", s / 10),
        }
        return;
      }
      try {
        await ds.set_structure_as_default(this.structure_id as number);
        this.loadData();
        this.info = {
          show: true,
          text: t("messages.structure_now_is_current"),
        }
      } catch (error: any) {
        const res = parseAxiosError(error);
        this.info = {
          color: "error",
          show: true,
          timeout: 3000,
          text: Array.isArray(res) ? t(res[0], ...res.slice(1)) : t(res),
        }
        console.error(error);
      }
    },
    async update(item: StructureCategory) {
      try {
        await ds.update(this.structure_id as number, item);
        this.loadCategories();
        this.close();
        this.info = {
          show: true,
          text: t("messages.updated"),
        }
      } catch (error: any) {
        const res = parseAxiosError(error);
        this.info = {
          color: "error",
          show: true,
          timeout: 3000,
          text: Array.isArray(res) ? t(res[0], ...res.slice(1)) : t(res),
        }
        console.error(error);
      }
    },
  },
  mounted() {
    state.settings[pageKey] = loadParams(pageKey) || state.settings[pageKey];
    state.refreshCallbacks[pageKey] = this.loadData;
    this.loadData();
  },
});
</script>
