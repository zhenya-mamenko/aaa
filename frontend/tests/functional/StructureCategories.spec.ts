import { mountComponent, baseStubs } from "./utils";
import db from "../db_data.json"
// @ts-ignore
import StructureCategories from "@/components/StructureCategories.vue";


const structures = db.structures;
const categories = db.categories;

jest.mock("@/common/locale", () => ({
  t: (key: string) => key,
}));

jest.mock("@/services/StructuresDataService", () => ({
  getAll: (structure_id?: number) => Promise.resolve({ data: structure_id === undefined ? structures : categories, }),
}));


describe("StructureCategories", () => {

  function mount(data: any) {
    return mountComponent(
      StructureCategories,
      {
        simpleStubs: ["v-btn", "v-divider", "v-spacer", "form-dialog", "delete-confirmation-dialog"],
        stubs: [...baseStubs, ],
      },
      {},
      {
        categories,
        icon: "mdi-label-percent",
        is_current: false,
        structures,
        structure_id: 0,
        title: "Test StructureCategories",
        ...data,
      },
    );
  }

  it("Should render", () => {
    const wrapper = mount({});

    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Should contain the right text", () => {
    const wrapper = mount({});

    expect(wrapper.text()).toContain("Test StructureCategories");
    const headers = wrapper.findAll("th");
    expect(headers.length).toBe(3);
    expect(headers[0].text()).toBe("categories.1");
    expect(headers[1].text()).toBe("portfolio.target_percent");
    expect(headers[2].text()).toBe("actions");
  });

  it("Should contain the right data", () => {
    const wrapper = mount({});

    const rows = wrapper.findAll("tbody tr");
    expect(rows.length).toBe(4);

    const data = rows[0].findAll("td");
    expect(data.length).toBe(3);
    expect(data[0].text()).toBe("Category 1");
    expect(data[1].text()).toBe("10%");
    expect(data[2].text()).toContain("mdi-pencil");
    expect(data[2].text()).toContain("mdi-delete");
  });

});
