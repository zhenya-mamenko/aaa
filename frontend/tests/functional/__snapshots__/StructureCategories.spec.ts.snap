// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StructureCategories Should render 1`] = `
"<div class="v-table v-table--has-top v-table--has-bottom v-theme--light v-table--density-default v-data-table h-100" data-testid="structure-categories-data-table">
  <div flat="">
    <div class="ms-3 me-3" size="large">mdi-label-percent</div>
    <div class="d-none d-md-block ms-0">Test StructureCategories</div>
    <div class="w-25">
      <div class="v-input v-input--horizontal v-input--center-affix v-input--density-compact v-theme--light v-locale--is-ltr v-input--dirty v-text-field v-select v-select--single v-select--selected me-2" data-testid="structure-categories-structure-select">
        <!---->
        <div class="v-input__control">
          <div class="v-field v-field--active v-field--appended v-field--center-affix v-field--dirty v-field--prepended v-field--no-label v-field--variant-filled v-theme--light v-locale--is-ltr" role="combobox">
            <div class="v-field__overlay"></div>
            <div class="v-field__loader">
              <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
                <!---->
                <div class="v-progress-linear__background"></div>
                <div class="v-progress-linear__buffer" style="width: 0%;"></div>
                <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
                  <div class="v-progress-linear__indeterminate">
                    <div class="v-progress-linear__indeterminate long"></div>
                    <div class="v-progress-linear__indeterminate short"></div>
                  </div>
                </transition-stub>
                <!---->
              </div>
            </div>
            <div class="v-field__prepend-inner">
              <div icon="mdi-list-box"></div>
              <!---->
            </div>
            <div class="v-field__field" data-no-activator="">
              <!----><label class="v-label v-field-label" for="input-0">
                <!---->
                <!---->
              </label>
              <!---->
              <div class="v-field__input" data-no-activator="">
                <!---->
                <!---->
                <div class="v-select__selection"><span class="v-select__selection-text">0<!----></span></div><input size="1" type="text" id="input-0" aria-describedby="input-0-messages" inputmode="none" aria-label="Open" title="Open" value="0">
              </div>
              <!---->
            </div>
            <!---->
            <div class="v-field__append-inner">
              <!---->
              <div class="v-select__menu-icon" icon="$dropdown"></div>
              <!---->
            </div>
            <div class="v-field__outline">
              <!---->
              <!---->
            </div>
          </div>
        </div>
        <!---->
        <!---->
      </div>
    </div>
    <v-divider-stub class="ms-3 me-3" vertical=""></v-divider-stub>
    <v-btn-stub class="me-2" data-testid="structure-categories-add-button" color="primary-darken-3" disabled="false"></v-btn-stub>
  </div>
  <div class="v-table__wrapper">
    <table>
      <!---->
      <thead>
        <tr>
          <th class="v-data-table__td v-data-table-column--align-start v-data-table__th--sortable v-data-table__th v-data-table__th--sticky" style="position: sticky;" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>categories.1</span>
              <div class="v-data-table-header__sort-icon" icon="mdi-triangle-small-up"></div>
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-end v-data-table__th--sortable v-data-table__th v-data-table__th--sticky" style="width: 100px; max-width: 130px; position: sticky;" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>portfolio.target_percent</span>
              <div class="v-data-table-header__sort-icon" icon="mdi-triangle-small-up"></div>
              <!---->
            </div>
          </th>
          <th class="v-data-table__td v-data-table-column--align-center v-data-table__th v-data-table__th--sticky" style="width: 80px; max-width: 80px; position: sticky;" colspan="1" rowspan="1">
            <div class="v-data-table-header__content"><span>actions</span>
              <!---->
              <!---->
            </div>
          </th>
        </tr>
        <!---->
      </thead>
      <!---->
      <tbody>
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">Category 1</td>
          <td class="v-data-table__td v-data-table-column--align-end" style="width: 100px; max-width: 130px;">10%</td>
          <td class="v-data-table__td v-data-table-column--align-center" style="width: 80px; max-width: 80px;">
            <div class="me-2" size="small"> mdi-pencil </div>
            <div size="small"> mdi-delete </div>
          </td>
        </tr>
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">Category 2</td>
          <td class="v-data-table__td v-data-table-column--align-end" style="width: 100px; max-width: 130px;">20%</td>
          <td class="v-data-table__td v-data-table-column--align-center" style="width: 80px; max-width: 80px;">
            <div class="me-2" size="small"> mdi-pencil </div>
            <div size="small"> mdi-delete </div>
          </td>
        </tr>
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">Category 3</td>
          <td class="v-data-table__td v-data-table-column--align-end" style="width: 100px; max-width: 130px;">30%</td>
          <td class="v-data-table__td v-data-table-column--align-center" style="width: 80px; max-width: 80px;">
            <div class="me-2" size="small"> mdi-pencil </div>
            <div size="small"> mdi-delete </div>
          </td>
        </tr>
        <!---->
        <tr class="v-data-table__tr">
          <td class="v-data-table__td v-data-table-column--align-start">Category 4</td>
          <td class="v-data-table__td v-data-table-column--align-end" style="width: 100px; max-width: 130px;">40%</td>
          <td class="v-data-table__td v-data-table-column--align-center" style="width: 80px; max-width: 80px;">
            <div class="me-2" size="small"> mdi-pencil </div>
            <div size="small"> mdi-delete </div>
          </td>
        </tr>
        <!---->
        <!---->
      </tbody>
      <!---->
      <!---->
    </table>
  </div>
  <div flat="">
    <v-btn-stub class="ms-3" data-testid="structure-categories-set-as-current-button" color="primary-darken-3" disabled="false"></v-btn-stub>
    <v-divider-stub class="ms-3 me-3" vertical=""></v-divider-stub>
    <v-spacer-stub></v-spacer-stub>
    <div class="text-right text-h6 me-3">0% <div class="ms-1" color="error" icon="mdi-alert" size="small"></div>
    </div>
  </div>
</div>
<form-dialog-stub opened="false" params="[object Object]" show-reset-button="false" title=""></form-dialog-stub>
<delete-confirmation-dialog-stub opened="false" text=""></delete-confirmation-dialog-stub>"
`;
